servicetalk-http-netty/src/test/java/io/servicetalk/http/netty/MalformedDataAfterHttpMessageTest.java

```diff
@@ -118,11 +123,20 @@ void afterResponse() throws Exception {
         }
     }
 
-    @Test
-    void afterResponseNextClientRequestSucceeds() throws Exception {
+    @ParameterizedTest
+    @ValueSource(booleans = {true, false})
+    void afterResponseNextClientRequestSucceeds(boolean doOffloading) throws Exception {
         Queue<ConnectionContext> contextQueue = new ArrayBlockingQueue<>(4);
         ServerSocketChannel server = nettyServer(RESPONSE_MSG);
         try (BlockingHttpClient client = stClientBuilder(server.localAddress())
+                .executionStrategy(doOffloading ? defaultStrategy() : noOffloadsStrategy())
+                // ClosedChannelException maybe observed on the second request if write is done before read of the
+                // garbage data, which won't be a RetryableExcepton. We may also see an exception from flush if the read
+                // closed the connection and then attempt to write on the same connection.
+                .appendClientFilter(new RetryingHttpRequesterFilter.Builder()
```

Comment:
@idelpivnitskiy - inserting retries kinda goes against the spirit of what we originally wanted to verify, but I don't think it can be done reliably with out a retry in place. wdyt?

Reply:
Interesting, thanks for looking at it. Agreed, the behavior can be racy and it's expected that we may race between connection selection and closure on the IO thread. We can keep as you proposed, or sequence requests without the race, waiting for the ctx1 to be closed:
```patch
Index: servicetalk-http-netty/src/test/java/io/servicetalk/http/netty/MalformedDataAfterHttpMessageTest.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/servicetalk-http-netty/src/test/java/io/servicetalk/http/netty/MalformedDataAfterHttpMessageTest.java b/servicetalk-http-netty/src/test/java/io/servicetalk/http/netty/MalformedDataAfterHttpMessageTest.java
--- a/servicetalk-http-netty/src/test/java/io/servicetalk/http/netty/MalformedDataAfterHttpMessageTest.java	(revision 58cb933d24f2c71d6cad99bb091b89fb80364f84)
+++ b/servicetalk-http-netty/src/test/java/io/servicetalk/http/netty/MalformedDataAfterHttpMessageTest.java	(date 1635803916686)
@@ -26,7 +26,6 @@
 import io.servicetalk.http.api.StreamingHttpConnectionFilter;
 import io.servicetalk.http.api.StreamingHttpRequest;
 import io.servicetalk.http.api.StreamingHttpResponse;
-import io.servicetalk.http.utils.RetryingHttpRequesterFilter;
 import io.servicetalk.transport.api.ConnectionContext;
 import io.servicetalk.transport.api.HostAndPort;
 import io.servicetalk.transport.api.ServerContext;
@@ -45,8 +44,6 @@
 import io.netty.handler.codec.http.HttpRequestDecoder;
 import org.junit.jupiter.api.Test;
 import org.junit.jupiter.api.extension.RegisterExtension;
-import org.junit.jupiter.params.ParameterizedTest;
-import org.junit.jupiter.params.provider.ValueSource;
 
 import java.io.IOException;
 import java.net.InetSocketAddress;
@@ -60,7 +57,6 @@
 import static io.netty.util.ReferenceCountUtil.release;
 import static io.servicetalk.buffer.api.Matchers.contentEqualTo;
 import static io.servicetalk.http.api.HttpExecutionStrategies.defaultStrategy;
-import static io.servicetalk.http.api.HttpExecutionStrategies.noOffloadsStrategy;
 import static io.servicetalk.http.api.HttpHeaderNames.CONTENT_LENGTH;
 import static io.servicetalk.http.api.HttpHeaderNames.CONTENT_TYPE;
 import static io.servicetalk.http.api.HttpHeaderValues.TEXT_PLAIN;
@@ -75,7 +71,6 @@
 import static io.servicetalk.transport.netty.internal.EventLoopAwareNettyIoExecutors.toEventLoopAwareNettyIoExecutor;
 import static java.lang.Integer.MAX_VALUE;
 import static java.lang.String.valueOf;
-import static java.time.Duration.ofNanos;
 import static org.hamcrest.MatcherAssert.assertThat;
 import static org.hamcrest.Matchers.anyOf;
 import static org.hamcrest.Matchers.empty;
@@ -123,20 +118,11 @@
         }
     }
 
-    @ParameterizedTest
-    @ValueSource(booleans = {true, false})
-    void afterResponseNextClientRequestSucceeds(boolean doOffloading) throws Exception {
+    @Test
+    void afterResponseNextClientRequestSucceeds() throws Exception {
         Queue<ConnectionContext> contextQueue = new ArrayBlockingQueue<>(4);
         ServerSocketChannel server = nettyServer(RESPONSE_MSG);
         try (BlockingHttpClient client = stClientBuilder(server.localAddress())
-                .executionStrategy(doOffloading ? defaultStrategy() : noOffloadsStrategy())
-                // ClosedChannelException maybe observed on the second request if write is done before read of the
-                // garbage data, which won't be a RetryableExcepton. We may also see an exception from flush if the read
-                // closed the connection and then attempt to write on the same connection.
-                .appendClientFilter(new RetryingHttpRequesterFilter.Builder()
-                        .maxRetries(MAX_VALUE)
-                        .retryFor((req, cause) -> true)
-                        .buildWithConstantBackoffFullJitter(ofNanos(1)))
                 .appendConnectionFilter(connection -> new StreamingHttpConnectionFilter(connection) {
                     @Override
                     public Single<StreamingHttpResponse> request(final HttpExecutionStrategy strategy,
@@ -147,19 +133,15 @@
                 })
                 .buildBlocking()) {
             validateClientResponse(client.request(client.get("/1")));
-            validateClientResponse(client.request(client.get("/2")));
-
             ConnectionContext ctx1 = contextQueue.poll();
             assertThat(ctx1, not(nullValue()));
-            // RetryingHttpRequesterFilter or AutoRetry may re-issue the request if a failure is seen locally. Verify
-            // the last connection (used for second request) is different from the first.
-            ConnectionContext ctx2 = null;
-            ConnectionContext tmp;
-            while ((tmp = contextQueue.poll()) != null) {
-                ctx2 = tmp;
-            }
+            ctx1.onClose().toFuture().get();
+
+            validateClientResponse(client.request(client.get("/2")));
+            ConnectionContext ctx2 = contextQueue.poll();
             assertThat(ctx2, not(nullValue()));
-            assertThat(ctx1, not(equalTo(ctx2)));
+            assertThat(ctx2, not(equalTo(ctx1)));
+            ctx2.onClose().toFuture().get();
             assertThat(contextQueue, empty());
         } finally {
             server.close().sync();
```

Reply:
the read may not happen until the next request is issued so I'll stick with the retry approach for now.